src/batchnorm.jl

```diff
@@ -0,0 +1,485 @@
+"""
+
+`BNMoments` is a high-level data structure used to store running mean and running variance
+of batch normalization.
+
+# Fields
+ `momentum::AbstractFloat`: A real number between 0 and 1 to be used as the scale of
+  last mean and variance. The existing running mean or variance is multiplied with 
+  (1-momentum).
+ 
+ `mean`: The running mean.
+
+ `var`: The running variance.
+ 
+ `meaninit`: The function used for initialize the running mean. Should either be `nothing` or
+of the form `(eltype, dims...) -> data`. `zeros` is a good option.
+
+ `varinit`: The function used for initialize the running variance. Should either be `nothing` or
+`(eltype, dims...) -> data`. `ones` is a good option.
+
+# Constructors
+ `BNMoments(;momentum=0.1, meaninit=zeros, varinit=ones)` sets `mean` and `var` to nothing.
+When this constructor is used, `batchnorm` functions will initialize `mean` and `var` during
+the first training iteration to have relevant dimensions and match the data type of the input.
+
+ `BNMoments(momentum, mean, var)` can be used directly load moments from data. `meaninit` and
```

Comment:
OK, so if I don't know anything about mean/var I just use bnmoments() before training to initialize. It will learn the real mean/var.  During testing it will stop updating itself and use the final frozen values.  I only need bnmoments(momentum, mean, var) if I am doing something special and loading mean/var from external source.  Is this during training or testing?  If testing, why do I need momentum?  If training what is the difference between this and starting bnmoments with a different init?

Reply:
I realized forcing to provide `momentum` when giving `mean` and `var` is problematic when loading pre-trained models just for inference. `meaninit` and `varinit` are only called when `mean` and `var` are nothing, so we may reduce `bnmoments` to a single function with default arguments.  

Calling `bnmoments(mean, var)` and `bnmoments(meaninit=(_)->mean, varinit=(_)->var)` are similar; but calling the second will provide automatic type conversion (for instance, if `mean` is an `Array` and `x` is  a`KnetArray`, mean will be made `KnetArray` and stored as a `KnetArray` in the `moments`). This behaviour is provided by the `_lazy_init!` helper function and works when `x` is a `Rec` or first call made when `training=false`. When `mean` and `var` are provided by the user, they are not further processed and used directly.